### DISCUSSION TITLE
Should `BAD_USER_INPUT` raise a 400 error?

### DISCUSSION THREAD

**Discussion by IodizedGabe** [apollographql/apollo-server#7460] [General] [answered]
According to the [docs](https://www.apollographql.com/docs/apollo-server/data/errors/#setting-http-status-code-and-headers), a failure to parse/validate a request should raise a 400.

This should imply that sending a request with required variables and no input should also raise a 400, because the request as a whole is technically invalid:
```
query ItemById($itemId: ID!) {
    item(itemId: $itemId) {
        name
    }
}

Input: {}
```

However, when upgrading from `apollo-server-core`: 3.11 => `@apollo/server`: 4.5.0, I noticed that this same request started returning 200 instead. Was this an intentional change, or a regression?

**Comment by trevor-scheer** [comment, marked as answer]
Fair question, there's maybe some nuance to the language here. Assuming you're referring to:
>If Apollo Server can't parse the request into a legal GraphQL document and validate it against your schema, it responds with a 400 status code
The parse and validate here refer specifically to the document - does it parse (is it valid graphql) and does it validate _against the schema_ (can we execute it).

Invalid _variables_ actually surface during execution, mostly as an unfortunate detail of `graphql-js`'s variable resolution being baked into execution (described at length here): https://github.com/apollographql/apollo-server/blob/d7e9b97595b063f1e796ec4449850a16d19e8b18/packages/server/src/requestPipeline.ts#L447-L468.

This case is falling awkwardly into the last bullet, and I can sympathize with any confusion on this one:
>Otherwise, Apollo Server returns a 200 status code. This is essentially the case where the server can execute the GraphQL operation, and execution completes successfully (though this can still include resolver-specific errors)
...we should probably update that to include variable resolution failure to be more clear.

If that status code is important to you, you can configure it with a plugin. Take a look at the examples in the docs [here](https://www.apollographql.com/docs/apollo-server/data/errors#setting-http-status-code-and-headers) and let me know if that's helpful.

**Reply by IodizedGabe** [reply]
Thanks for the confirmation.

Should this be raised as an error with graphql-js, then? Or is this a case of ambiguity in the spec?

**Reply by jessemyers** [reply]
> you can configure it with a plugin

While this is true, doing so in our project would result in multiple configuration elements for error handling because we also need to implement a custom `formatError` function. 

My take is that the current behavior is confusing and unnecessarily difficult to customize; would there be willingness to accept contributions that reduce the customization overhead?

**Reply by glasser** [reply]
I think @jessemyers and @IodizedGabe are right and this was an unintentional v4 regression that I introduced. (Whether the correct answer is "UserInputError should be a 400" or "the case where `execute` produces no `data` field (which is roughly the same as where AS proactively creates a UserInputError) should be a 400" is slightly less clear to me.)  More at https://github.com/apollographql/apollo-server/issues/7462#issuecomment-1481675591

**Reply by trevor-scheer** [reply]
Sorry for the confusion here, had to hash this one out with @glasser. Opened a PR to provide an opt-in mitigation here: https://github.com/apollographql/apollo-server/pull/7465
